File: java/src/test/java/gameoflife/GridManager.java
package gameoflife;

public class GridManager {

    public Cell[][] getNewState(Cell[][] grid) {
        Cell[][] newGrid = new Cell[grid.length][grid[0].length];
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                newGrid[i][j] = new Cell(false);
            }
        }

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                int nbNeighbours = countNeighbours(grid, i, j);
                if (nbNeighbours == 2) {
                    newGrid[i][j].isAlive = grid[i][j].isAlive();
                } else if (nbNeighbours == 3) {
                    newGrid[i][j].isAlive = true;
                }
            }
        }
        return newGrid;
    }

    public int countNeighbours(Cell[][] grid, int i, int j) {
        int nbNeighbours = 0;
        if (IsAliveNeighbour(grid, i - 1, j - 1)) {
            nbNeighbours++;
        }
        if (IsAliveNeighbour(grid, i, j - 1)) {
            nbNeighbours++;
        }
        if (IsAliveNeighbour(grid, i + 1, j - 1)) {
            nbNeighbours++;
        }
        if (IsAliveNeighbour(grid, i - 1, j)) {
            nbNeighbours++;
        }
        if (IsAliveNeighbour(grid, i + 1, j)) {
            nbNeighbours++;
        }
        if (IsAliveNeighbour(grid, i - 1, j + 1)) {
            nbNeighbours++;
        }
        if (IsAliveNeighbour(grid, i, j + 1)) {
            nbNeighbours++;
        }
        if (IsAliveNeighbour(grid, i + 1, j + 1)) {
            nbNeighbours++;
        }

        return nbNeighbours;
    }

    private boolean IsAliveNeighbour(Cell[][] grid, int i, int j) {
        return j >= 0 &&
                j < grid[0].length &&
                i >= 0 &&
                i < grid.length &&
                grid[i][j].isAlive();
    }


}

File: Python/game_of_life/grid_manager.py
from game_of_life.cell import Cell


class GridManager(object):
    def get_new_state(self, grid):
        new_grid = [[Cell(False) for _ in range(len(grid[0]))] for _ in range(len(grid))]

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                nb_neighbours = self.count_neighbours(grid, i, j)
                if nb_neighbours == 2:
                    new_grid[i][j].is_alive = grid[i][j].is_alive
                elif nb_neighbours == 3:
                    new_grid[i][j].is_alive = True

        return new_grid

    def count_neighbours(self, grid, i, j):
        nb_neighbours = 0
        if self._is_alive_neighbour(grid, i - 1, j - 1):
            nb_neighbours += 1
        if self._is_alive_neighbour(grid, i, j - 1):
            nb_neighbours += 1
        if self._is_alive_neighbour(grid, i + 1, j - 1):
            nb_neighbours += 1
        if self._is_alive_neighbour(grid, i - 1, j):
            nb_neighbours += 1
        if self._is_alive_neighbour(grid, i + 1, j):
            nb_neighbours += 1
        if self._is_alive_neighbour(grid, i - 1, j + 1):
            nb_neighbours += 1
        if self._is_alive_neighbour(grid, i, j + 1):
            nb_neighbours += 1
        if self._is_alive_neighbour(grid, i + 1, j + 1):
            nb_neighbours += 1
        return nb_neighbours

    @staticmethod
    def _is_alive_neighbour(grid, i, j):
        return 0 <= i < len(grid) and \
               0 <= j < len(grid[0]) and \
               grid[i][j].is_alive
